Refactor action testing

markscheel opened the pull request:
## Proposed changes

Splits ActionTesting.hpp into 4 files, and renames some classes and variables.

This looks like a large PR, but it is mostly moving code around and should be easy to review; see comments in each commit message.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

markscheel committed:
Split ActionTesting.hpp into four files.

No actual changes to any code, except for moving blocks of code from
one file to another, clang-formatting, and changes in includes and
forward declarations that were necessary when code blocks were moved.

New files:
 - MockRuntimeSystem.hpp contains the MockRuntimeSystem class.
 - MockRuntimeSystemFreeFunctions.hpp contains the free functions
   that are used to invoke member functions of MockRuntimeSystem
   and a few other free functions that depend on MockRuntimeSystem.
 - MockDistributedObject.hpp contains the MockDistributedObject class.

markscheel committed:
Test_EventsAndTriggers: Use free functions for methods on runner.

nilsdeppe commented on tests/Unit/Framework/ActionTesting.hpp:
```diff
@@ -391,9 +390,9 @@ class MockArrayElementProxy {
+  MockArrayElementProxy(MockChareElement<Component>& mock_chare_element,
                         Inbox& inbox)
-      : local_algorithm_(local_algorithm), inbox_(inbox) {}
+      : mock_chare_element_(mock_chare_element), inbox_(inbox) {}
```
Would `mock_distributed_objects` be clearer? I personally don't really know what `mock_chare_element` is and I'm worried that `element` will be confused as a DG element. Or since this is all the distributed objects, `all_mock_distributed_objects`?

nilsdeppe commented on tests/Unit/Framework/ActionTesting.hpp:
```diff
@@ -408,35 +407,35 @@ class MockArrayElementProxy {
   void perform_algorithm(const bool /*restart_if_terminated*/) noexcept {}
 
-  MockDistributedObject<Component>* ckLocal() { return &local_algorithm_; }
+  MockChareElement<Component>* ckLocal() { return &mock_chare_element_; }
```
I find that I have no clue what a `MockChareElement` is. A chare element isn't even a concept in Charm++ so that seems like a bad name. Why is `MockDistributedObject` bad? This is exactly what it is.

markscheel replied:
I'm happy to discuss better names. 

`MockChareElement` is the thing referred to by a `MockArrayElementProxy`, i.e. a single element of an Array chare, Group chare, or NodeGroup chare.  `MockChareElement` does not represent an entire chare (even for singletons which we treat as an Array with one element). I added 'Element' to the name `MockChareElement` to match 'Element' in the name `MockArrayElementProxy`.  (As opposed to `MockProxy` which is a proxy to an entire chare).

I am also considering renaming `MockArrayElementProxy` -> `MockChareElementProxy` to get the names further in sync, and reflect the fact that we will soon have Groups and NodeGroups too.

I find the name `MockDistributedObject` very confusing, because the object is not "distributed": it is one element of an array chare that lives on a single core somewhere.   When I see the name `MockDistributedObject` I think of an entire array chare, which is distributed over many cores/nodes (like the thing that `MockProxy` refers to, which is an unordered_map of `MockChareElement`s aka `MockDistributedObject`s that is indexed by an `array_index`). Much of my confusion in ActionTesting came from the name `MockDistributedObject` (and that objects of type `MockDistributedObject` were named `local_algorithm`, and I didn't understand how something can be both distributed and local).

markscheel replied:
Agreed at the SpECTRE call Dec 9 to keep it MockDistributedObject, and to eventually rename AlgorithmImpl -> DistributedObject.  "Distributed Object" is a regular term in HPC meaning an object that can be called remotely.

markscheel committed:
ActionTesting: rename some classes and variables.

MockArrayElementProxy -> MockDistributedObjectProxy
MockProxy->MockCollectionOfDistributedObjectsProxy
TupleOfMockDistributedObjects->CollectionOfMockDistributedObjects

And now objects of type MockDistributedObject are named
mock_distributed_object, and so forth.

Also added more documentation of the name MockDistributedObject.

markscheel commented:
Changed names as (i think) agreed on in the SpECTRE meeting.

nilsdeppe reviewed (approved):
Well I was going to file #2639 but you beat me too it :D 

I find the documentation is clear and correct :) Thank you!
